compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -596,8 +596,13 @@ object Semantic {
         && meth.owner.is(Flags.Module)
         && meth.owner.companionClass.is(Flags.Case)
 
+      def isAlwaysSafe(meth: Symbol) =
+        Set("eq", "ne", "isInstanceOf", "asInstanceOf").contains(meth.name.toString())
```

Comment:
Use `meth eq Definitions.Object_eq` etc. This ensures that the method really is *the* `eq` method on the object class, and does not match a user-defined method whose name just happens to be `eq`.